Sources/Sentry/SentryHub.m

```diff
@@ -9,66 +9,94 @@
 #if __has_include(<Sentry/Sentry.h>)
 #import <Sentry/SentryHub.h>
 #import <Sentry/SentryClient.h>
-#import <Sentry/SentryBreadcrumbStore.h>
+#import <Sentry/SentryStackLayer.h>
+#import <Sentry/SentryBreadcrumbTracker.h>
+#import <Sentry/SentryBreadcrumbs.h>
 #else
 #import "SentryHub.h"
 #import "SentryClient.h"
-#import "SentryBreadcrumbStore.h"
+#import "SentrySentryStackLayer.h"
+#import "SentryBreadcrumbTracker.h"
+#import "SentryBreadcrumbs.h"
 #endif
 
 @interface SentryHub()
 
-@property (nonatomic, strong) SentryClient *client;
+@property (nonatomic, strong) NSMutableArray<SentryStackLayer *> *stack;
 
 @end
 
 @implementation SentryHub
 
-+ (SentryHub *)defaultHub {
-    static SentryHub *_sharedInstance = nil;
-    static dispatch_once_t oncePredicate;
-    dispatch_once(&oncePredicate, ^{
-        _sharedInstance = [[self alloc] init];
-    });
-    return _sharedInstance;
-}
-
-- (instancetype)initWithClient:(SentryClient *)aClient {
-    self = [super init];
-    if (self) {
-        [self bindClient:aClient];
+- (instancetype)init {
+    if (self = [super init]) {
+        SentryScope *scope = [[SentryScope alloc] init];
+        SentryStackLayer *layer = [[SentryStackLayer alloc] init];
+        layer.scope = scope;
+        [self setStack:[@[layer] mutableCopy]];
     }
     return self;
 }
+/**
+ */
+- (void)setupWithClient:(SentryClient * _Nullable)client {
+    SentryStackLayer *stackLayer = [SentryStackLayer new];
+    stackLayer.scope = [SentryScope new];
+    [self setStack:[@[stackLayer] mutableCopy]];
+}
 
 - (void)captureEvent:(SentryEvent *)event {
-    [self.client sendEvent:event withCompletionHandler:nil];
+    [[self getClient] sendEvent:event scope:[self getScope] withCompletionHandler:nil];
 }
 
 - (void)addBreadcrumb:(SentryBreadcrumb *)crumb {
-    [self.client.breadcrumbs addBreadcrumb:crumb];
+    [[self getScope].breadcrumbs addBreadcrumb:crumb];
 }
 
 - (SentryClient * _Nullable)getClient {
-    return self.client;
+    if (nil != [self getStackTop]) {
+        return [[self getStackTop] client];
+    }
+    return nil;
+}
+
+- (void)bindClient:(SentryClient * _Nullable)client {
+    if (nil != [self getStackTop]) {
+        [self getStackTop].client = client;
+    }
+}
+
+- (SentryStackLayer *)getStackTop {
+    return self.stack[self.stack.count - 1];
 }
 
-- (void)bindClient:(SentryClient *)aClient {
-    [self setClient:aClient];
+- (SentryScope *)getScope {
+    return [self getStackTop].scope;
+}
 
-    // TODO(fetzig): remove this as soon as SentryHub is fully capable of managing multiple `SentryClient`s
-    [SentryClient setSharedClient:aClient];
+- (SentryScope *)pushScope {
+    SentryScope * scope = [[[self getStackTop] scope] copy];
+    SentryStackLayer *newLayer = [SentryStackLayer new];
+    newLayer.scope = scope;
```

Comment:
Don't we also need to set the `client` in the new layer?
We can just set the same client for now since cloning it isn't easily possible right now but add a todo here once the client is fully ported :)

Reply:
*facepalm*